### PR TITLE
Change return type of TemplateParameters' methods from Option to just…

### LINKED ISSUE #960: TemplateParameters' methods should just return a Vec, not an Option<Vec>
I thought that the `Option` would be useful, but in practice we do pretty much ignore that it is an option and do things like `params.unwrap_or(vec![])`. And then in other places when the result is `Some` we assume that the contained `Vec` is not empty, when that isn't always the case.

It will be much simpler to just return a `Vec` that is either empty or not.

And also `usize` instead of `Option<usize>` for `num_self_template_params`.

The trait's comment will need to be updated as well.

https://github.com/rust-lang-nursery/rust-bindgen/blob/cd41e5cfe26555af574e81b6aa3bed2b84a18000/src/ir/template.rs#L102

### PR DESCRIPTION
**Pull request by hgallagher1993**
… Vec

Fixes #960

### CODE REVIEW

**Comment by emilio** [comment]
Seems like this broke a bunch of tests...

**Comment by fitzgen** [comment]
Hi @hgallagher1993, did you figure out how to run the tests locally? It looks like this PR changed a lot of the bindings generated for the test headers, which I wouldn't expect.

**Comment by hgallagher1993** [comment]
@fitzgen I had a read through the contributing.md and do mean to just run `cargo test`?

**Comment by hgallagher1993** [comment]
I ran `cargo test` and the differences seem to be pretty much just white space that changed, in `16-byte-alignment.rs` for example 

```
fn bindgen_test_layout_rte_ipv4_tuple__bindgen_ty_1__bindgen_ty_1() {

    assert_eq!(::std::mem::size_of::<rte_ipv4_tuple__bindgen_ty_1__bindgen_ty_1>()

               , 4usize , concat ! (

               "Size of: " , stringify ! (

               rte_ipv4_tuple__bindgen_ty_1__bindgen_ty_1 ) ));

```
changed to

```
fn bindgen_test_layout_rte_ipv4_tuple__bindgen_ty_1__bindgen_ty_1() {

    assert_eq!(

        ::std::mem::size_of::<rte_ipv4_tuple__bindgen_ty_1__bindgen_ty_1>(),

        4usize,

        concat!(

            "Size of: ",

            stringify!(rte_ipv4_tuple__bindgen_ty_1__bindgen_ty_1)

        )

    );
```

And all the `impl Clone. . .` dissappeared and went in to the `#[derive. . .]`'s

**Comment by fitzgen** [comment]
@hgallagher1993 yes. Make sure you have the latest nightly and `cargo +nightly install -f rustfmt-nightly`. That should take care of the formatting issues.

The `impl Clone` moving to a `derive` is more worrying; that means that this PR changed something to do with how we decide whether to derive or implement `Clone`, which it probably shouldn't.

This is what `src/codegen/mod.rs` does before this PR:

```rust
        if item.can_derive_copy(ctx) && !item.annotations().disallow_copy() &&
            ctx.options().derive_copy
        {
            derives.push("Copy");
            if used_template_params.is_some() {
                // FIXME: This requires extra logic if you have a big array in a
                // templated struct. The reason for this is that the magic:
                //     fn clone(&self) -> Self { *self }
                // doesn't work for templates.
                //
                // It's not hard to fix though.
                derives.push("Clone");
            } else {
                needs_clone_impl = true;
            }
        }
```

Were we really getting `Some(v)` where `v.is_empty()` before with the tests that changed? Or did we subtly change the logic in other ways?

File: src/codegen/mod.rs [hunk 1]
```diff
@@ -1450,7 +1449,7 @@ impl CodeGenerator for CompInfo {
             ctx.options().derive_copy
         {
             derives.push("Copy");
-            if used_template_params.is_some() {
+            if used_template_params.is_empty() {
```
Comment (fitzgen) [inline src/codegen/mod.rs]: I guess we must have been hitting `Some(v)` where `v.is_empty()` fairly often.

Comment (hgallagher1993) [reply src/codegen/mod.rs]: Should this be `!used_template_params.is_empty()`? When the ! is put in the `impl clone`'s are left alone when the tests are run.

Comment (fitzgen) [reply src/codegen/mod.rs]: Totally. Should have caught this earlier! Good eye ;)

**Comment by hgallagher1993** [comment]
@fitzgen I've ran the tests and added in the `!`. . .there's still the merge conflict to go though

**Comment by fitzgen** [comment]
Great! Let me know when you've rebased (or if you need any help rebasing) and then I'll take a final look and we can merge this thing :)

**Comment by hgallagher1993** [comment]
@fitzgen The merge conflict should be gone now

**Comment by hgallagher1993** [comment]
I ran the tests locally and they all passed so I'm not really sure why it's failed here

**Comment by pepyakin** [comment]
@hgallagher1993 
Hello! According to the travis build log `cargo test` [works fine](https://travis-ci.org/rust-lang-nursery/rust-bindgen/jobs/289174807) on LLVM 3.8 and fails on [other](https://travis-ci.org/rust-lang-nursery/rust-bindgen/jobs/289174813) [versions](https://travis-ci.org/rust-lang-nursery/rust-bindgen/jobs/289174821). 
If we will take a look on the failing tests we will see these:

- partial-specialization-and-inheritance.hpp
- auto.hpp

These tests are special in the sense that they doesn't have the only one expected binding! 
So for example there are 3 different bindings for auto.hpp depending on LLVM version:

- [tests/expectations/tests/libclang-3.8/auto.rs](https://github.com/rust-lang-nursery/rust-bindgen/blob/master/tests/expectations/tests/libclang-3.8/auto.rs)
- [tests/expectations/tests/libclang-3.9/auto.rs](https://github.com/rust-lang-nursery/rust-bindgen/blob/master/tests/expectations/tests/libclang-3.9/auto.rs)
- [tests/expectations/tests/libclang-4/auto.rs](https://github.com/rust-lang-nursery/rust-bindgen/blob/master/tests/expectations/tests/libclang-4/auto.rs)

You can read [more here](https://github.com/rust-lang-nursery/rust-bindgen/blob/master/CONTRIBUTING.md#test-expectations-and-libclang-versions).

But unfortunatelly, this is not the only problem. If you take a look into [`expectations` test logs](https://travis-ci.org/rust-lang-nursery/rust-bindgen/jobs/289174811) you can see some failures. 
The `expectations` tests just compile every binding and run layout in it. 
Note that this isn't done when you issue `cargo test`.
You can run these tests with the following commands:

```
$ cd tests/expectations
$ cargo test
```

**Comment by hgallagher1993** [comment]
@pepyakin Ya it was only `cargo test` I ran and I can see the tests that failed locally now so I should have this fixed up in the next day or 2

**Comment by emilio** [comment]
Closing because this bitrotted a lot and is inactive, but thanks for the patch @hgallagher1993! :)

**Pull request closed without merging**